src/compiler/utilities.ts

```diff
@@ -2713,6 +2713,13 @@ namespace ts {
         return forEach(supportedTypeScriptExtensions, extension => fileExtensionIs(fileName, extension));
     }
 
+    /** Return ".ts" or ".tsx" if that is the extension. */
+    export function tryExtractTypeScriptExtension(fileName: string): string | undefined {
+        return tryFind(supportedTypescriptExtensionsWithDtsFirst, extension => fileExtensionIs(fileName, extension));
+    }
+    // Must have '.d.ts' first because if '.ts' goes first, that will be detected as the extension instead of '.d.ts'.
+    const supportedTypescriptExtensionsWithDtsFirst = supportedTypeScriptExtensions.slice().reverse();

```

Comment:
`supportedTypescriptExtensions` currently has a comment on it saying that it's in file resolution precedence. What else could we add?
I might want to change this code to just duplicate the list, though, because if anyone ever changes the order of `supportedTypescriptExtensions`, this one will no longer have `.d.ts` first.